docs/API_DESIGN.md

```diff
@@ -0,0 +1,658 @@
+
+  /// 🚧
+  PUBLIC DATA equivalentId: []string?
+
+  /// 🚧
+  PUBLIC DATA canonicalId: string?
+```
+
+## Methods
+
+### `DidJwk`
+
+```pseudocode!
+CLASS DidJwk
+  PUBLIC DATA did: Did
+  PUBLIC DATA document: Document
+  CONSTRUCTOR(public_key: Jwk)
+  CONSTRUCTOR(uri: string)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+#### Example: Create a `did:jwk`
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_jwk = new DidJwk(public_key)
+```
+
+#### Example: Instantiate an existing `did:jwk`
+
+```pseudocode!
+uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"
+did_jwk = new DidJwk(uri)
+```
+
+#### Example: Resolve a `did:jwk`
+
+```pseudocode!
+uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"
+resolution_result = DidJwk.resolve(uri)
+```
+
+### `DidWeb`
+
+```pseudocode!
+CLASS DidWeb
+  CONSTRUCTOR(uri: string)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+#### Example: Instantiate an existing `did:web`
+
+```pseudocode!
+uri = "did:web:example.com"
+did_web = new DidWeb(uri)
+```
+
+#### Example: Resolve a `did:web`
+
+```pseudocode!
+uri = "did:web:example.com"
+resolution_result = DidWeb.resolve(uri)
+```
+
+### `DidDht`
+
+```pseudocode!
+CLASS DidDht
+  PUBLIC DATA did: Did
+  PUBLIC DATA document: Document
+  CONSTRUCTOR(identity_key: Jwk)
+  CONSTRUCTOR(uri: string)
+  METHOD publish(signer: Signer)
+  METHOD deactivate(signer: Signer)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+> [!NOTE]
+> `resolve()` makes use of [`Ed25519Verifier`](#ed25519verifier) internally for DNS packet verification.
+
+#### Example: Create & publish a `did:dht`
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_dht = new DidDht(public_key)
+signer = key_manager.get_signer(public_key)
+did_dht.publish(signer)
+```
+
+#### Example: Create a `did:dht`, add to the Core Properties & publish
+
+> [!NOTE]
+> The call to the `new DidDht()` constructor only adds the minimum requirements to the DID Document.
+> If additional [Core Properties](https://www.w3.org/TR/did-core/#core-properties) are required, update the `document` data member prior-to the call to `publish()`.
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_dht = new DidDht(public_key)
+
+/// Set the alsoKnownAs
+did_dht.document.alsoKnownAs = "did:example:efgh"
```

Comment:
oof. this is not easy to do in all languages (we've already run into this issue in kotlin)

Reply:
You're referring to instantiating an instance and subsequently mutating it? What do you mean by "easy?"